src/charm.py

```diff
@@ -0,0 +1,222 @@
+        self.state = ClusterState(self, substrate=SUBSTRATE)
+        self.workload = ODWorkload()
+
+        # --- CHARM EVENT HANDLERS ---
+
+        self.password_action_events = PasswordActionEvents(self)
+        self.tls_events = TLSEvents(self)
+        self.requirer_events = RequirerEvents(self)
+        # self.upgrade_events = ODUpgradeEvents(
+        #     self,
+        #     dependency_model=OpensearchDasboardsDependencyModel(
+        #         **DEPENDENCIES  # pyright: ignore[reportGeneralTypeIssues]
+        #     ),
+        # )
+
+        # --- MANAGERS ---
+
+        self.tls_manager = TLSManager(
+            state=self.state, workload=self.workload, substrate=SUBSTRATE
+        )
+        self.config_manager = ConfigManager(
+            state=self.state, workload=self.workload, substrate=SUBSTRATE, config=self.config
+        )
+
+        # --- LIB EVENT HANDLERS ---
+
+        self.restart = RollingOpsManager(self, relation="restart", callback=self._restart)
+
+        # --- CORE EVENTS ---
+
+        self.framework.observe(getattr(self.on, "install"), self._on_install)
+
+        self.framework.observe(getattr(self.on, "start"), self._start)
+
+        self.framework.observe(
+            getattr(self.on, "update_status"), self._on_cluster_relation_changed
+        )
+        self.framework.observe(
+            getattr(self.on, "leader_elected"), self._on_cluster_relation_changed
+        )
+        self.framework.observe(
+            getattr(self.on, "config_changed"), self._on_cluster_relation_changed
+        )
+
+        self.framework.observe(
+            getattr(self.on, f"{PEER}_relation_changed"), self._on_cluster_relation_changed
+        )
+        self.framework.observe(
+            getattr(self.on, f"{PEER}_relation_joined"), self._on_cluster_relation_changed
+        )
+        self.framework.observe(
+            getattr(self.on, f"{PEER}_relation_departed"), self._on_cluster_relation_changed
+        )
+
+        self.framework.observe(getattr(self.on, "secret_changed"), self._on_secret_changed)
+
+    # --- CORE EVENT HANDLERS ---
+
+    def _on_install(self, event: InstallEvent) -> None:
+        """Handler for the `on_install` event."""
+        self.unit.status = MaintenanceStatus("installing Opensearch Dashboards...")
+
+        install = self.workload.install()
+        if not install:
+            self.unit.status = BlockedStatus("unable to install Opensearch Dashboards")
+
+        # don't complete install until passwords set
+        if not self.state.peer_relation:
+            self.unit.status = WaitingStatus("waiting for peer relation")
+            event.defer()
+            return
+
+        if self.unit.is_leader() and not self.state.cluster.internal_user_credentials:
+            for user in CHARM_USERS:
+                self.state.cluster.update({f"{user}-password": self.workload.generate_password()})
+
+    def _on_cluster_relation_changed(self, event: EventBase) -> None:
```

Comment:
nit: if we use `_on_cluster_relation_changed` for multiple events, then we could name it something more relevant, e.g. `reconcile`.

Reply:
We could....

@Mehdi-Bendriss what do you think?